### ISSUE TITLE
[Proposal] foolproof, auto-finalizing API design

### ISSUE THREAD

**Issue by jedwards1211** [archiverjs/node-archiver#458]
As I mentioned in #457, if you forget to call `.finalize()` (or accidentally delete the call for example), you'll see super-confusing behavior and possibly burn hours debugging it.  It would help everyone to have an API that prevents this risk.

Here's a different API design that prevents this problem entirely:

```js
const archive = archiver('zip', async archive => {
  archive.file('package.json', { name: `nodejs/package.json` })
  // I haven't found a way to make archive.glob put the entries in a subdirectory...
  const mainFiles = await promisify(glob)('**/*.js', { ignore: 'node_modules' })
  for (const file of mainFiles) archive.file(file, { name: `nodejs/node_modules/${file}` })
})
archive.pipe(fs.createWriteStream('test.zip'))
```

With this API you call all the methods to build the archive within this async builder function.  When the returned promise resolves, the archive finalizes automatically, so it's guaranteed to finalize (unless the userland code hangs).  If the returned promise rejects, the archive can emit an `error` event with the rejection reason.

Although promises are widely supported now, the builder function could also support callback style:

```js
const archive = archiver('zip', (archive, cb) => {
  archive.file('package.json', { name: `nodejs/package.json` })
  glob('**/*.js', { ignore: 'node_modules' }, (err, mainFiles) => {
    if (err) cb(err)
    else {
      for (const file of mainFiles) archive.file(file, { name: `nodejs/node_modules/${file}` })
      cb()
    }
  })
})
archive.pipe(fs.createWriteStream('test.zip'))
```

Of course, the risk of forgetting to call back is equivalent to forgetting to `.finalize()` with the current API, which is a general disadvantage of using callbacks instead of promises.

**Comment by ctalkington** [comment]
thanks for the suggestion! I have been wanting to go more modern and i think weve finally got our minimum required version to a place we can start exploring such

**Comment by jedwards1211** [comment]
Yeah, I can see how it wasn't reasonable back when you started the project
